### ISSUE TITLE
Tensor flow crashed using Adam optimizer

### ISSUE THREAD

**Issue by TristanPepin** [tensorforce/tensorforce#710]
Hi everyone, 

I'm currently using tensorforce for a project with @jerabaul29, and I noticed a crash from tensorflow during learning, due to Nan or inf values inside the NN. The error message is the following 

![image](https://user-images.githubusercontent.com/66601981/85835474-6de5ac80-b795-11ea-99f4-44b34ad79de2.png)

I'm using a ppo agent to perform the learning, with an adam optimizer, learning_rate = 1e-3. I'm also using a 2 layers dense neural network, with 512 neurons per layer. 

The Nan and Inf are not coming from my state or my reward function, so if anyone has any idea of what is going on, thanks in advance !

**Comment by jerabaul29** [comment]
Can it be that while your state has no Inf / NaN inside it, it has quite large values that saturate the network, and then if you are a bit unlucky with the coefficients of the ANN this gets even more amplified down the way?

Any idea on how to 'robustify' things further @AlexKuhnle ? I think that having some exceptions / warning popping out now and then should be ok, but what is a problem is that, if I understand well (can you confirm @TristanPepin ?), the resumed training crash in this case, probably due to some corrupted buffers somewhere....

**Comment by TristanPepin** [comment]
Yes perfectly, I forgot to mention that this bug occurred on approximately the 600th episode.  The resumed training crashed as well, and even if i load a previous but close checkpoint, the same crash happens few episodes later

**Comment by jerabaul29** [comment]
Ok, this definitely confirms some form of buffer contamination. This needs to be somehow fixed in tensorforce, as it is a very annoying behavior ^^ :) .

**Comment by AlexKuhnle** [comment]
Hi all,

First, just to check: is this on the latest master/pip version? And regarding your states and rewards, are they all within a reasonable range, not accidentally getting very large (even if not inf/nan)? Finally, can you post your agent and network config, to check whether something looks problematic.

Agreed that this behavior should be avoided. Introducing the inf/nan-check for gradients was a first step in that direction, same for input states nan/inf checks, but of course ideally the warning should happen as early as possible. Hard to say what could be a reason, maybe the weights are getting very big, maybe somewhere a division accidentally is not "safe". I hope it becomes clearer with the config.

**Comment by TristanPepin** [comment]
Hi,

Yes, this is on the latest master/pip version, and I just made a lot of tests and I found that in some very particular cases, a few values on my state vector were reaching almost 10**6 values (but rare), so it may be the explanation. I've corrected this, I am currently running a learning, I will let you know if this crash happens again. 

Here is my network configuration : 

`network = [dict(type='dense', size=512), dict(type='dense', size=512)] `

And the agent : 

`agent = Agent.create(
    # Agent + Environment
    agent='ppo', environment=environment, max_episode_timesteps=nb_state,
    # Network
    network=network,
    # Optimization
    batch_size=10, learning_rate=1e-3, subsampling_fraction=0.2, optimization_steps=25,
    # Reward estimation
    likelihood_ratio_clipping=0.2, estimate_terminal=True,  
    # Critic
    critic_network=network,
    critic_optimizer=dict(
        type='multi_step', num_steps=5,
        optimizer=dict(type='adam', learning_rate=1e-3)
    ),
    # Regularization
    entropy_regularization=0.01,
) `

**Comment by AlexKuhnle** [comment]
Would be interesting to see whether it's indeed the 10**6 values which cause problems -- while such values may not be great for learning, I wouldn't expect such values to cause actual inf/nan problems. Your config also looks quite "standard", nothing problematic standing out.

**Comment by jerabaul29** [comment]
I think that if you are unlucky the values could escalate quite quickly in a fully connected network, don t you think?

Something like:

- 1 input is 10^6
- if you are unlucky, at the 1st layer, you have 512 neurons that have values of typically 10^6
- if you are unlucky, at the second layer, you have 512 neurons with value 512 * 10^6
- if you are unlucky, at the output layer, you have an output of 512 * 512 * 10^6

combine that with gradient explosion in backprop, things explode quite quickly right? Also, is this float16 or 32 or 64 by default?

**Comment by AlexKuhnle** [comment]
It's float32 by default, which really should be okay even in the situation you describe. However, if that happens multiple times, weights might get too big and then infs/nans appear (still, there is [gradient clipping by default](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/optimizers/tf_optimizer.py#L79), so can't happen all that quickly). Also, just remember that dense layers use [tanh by default](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/layers/dense.py#L48), so again something which I think should prevent explosion.

Since you have the saved model, what might be worth having a look at is the size of the weights. You can access variables via `agent.get_variable(name)` and get all variable names via `agent.get_variables()`.

**Comment by TristanPepin** [comment]
The 10**6 (in fact sometimes more 10**15) singular values on my state vector seemed to be the problem, as I launched a new learning with a tight control of all vectors I'm feeding to my NN (max value = 100 no matter what) and i didn't have any problems during the learning !

**Comment by jerabaul29** [comment]
Ok, so this is fixed, right?

Then the conclusion is that there were some finite, but too big values in the state observation, right?

Should there be some checks and / or at least some serial print warnings in case any state / reward value is over 10^3 (or something higher) for example?

**Comment by TristanPepin** [comment]
Yeah it is, and yes, I think that this was the issues as the new learning had exactly the same parameters than the one that crashed, the only difference was this upper bound value for the state vector

**Comment by AlexKuhnle** [comment]
That makes sense. Yes, there should be a warning, I think, since even if technically possible to use such values, such magnitudes won't work well with any DRL algorithm (and even if it's rare outliers, training won't be bothered by clipping them). I'm currently thinking about how to improve this for the next version. Generally, the importance of bounded state spaces (and then normalizing the inputs given the bounds as a first preprocessing step) has been mentioned frequently in recent implementation-focused RL papers.

**Comment by AlexKuhnle** [comment]
I think this can be closed then, and warnings (or more likely changing defaults taking care of it) will be added in the next version.
